### PR TITLE
Change/uncharging strategy

### PR DESCRIPTION
**Pull request by karlthepagan**
please merge https://github.com/ScreepsOCS/screeps.behaviour-action-pattern/issues/471 first

workers will now pick from the closest container that puts them to work
haulers continue to pick from best-fit container
partially implemented minimum energy amount for managed containers

### CODE REVIEW

**Comment by karlthepagan** [comment]
I just noticed this won't have a big effect over RCL4 when you build storage.

If you like, you can flip uncharging & storage order in behaviour.worker.

File: creep.Action.js [hunk 1]
```diff
@@ -44,7 +44,24 @@ let Action = function(actionName){
     // find a new target for that action
     // needs implementation in derived action
     this.newTarget = function(creep){
-        return null;
+        const targetPool = creep.room.structures.container.in;
```
Comment (karlthepagan) [inline creep.Action.js]: removing this default. inappropriate targetPool

**Comment by karlthepagan** [comment]
rebased onto #471

File: creep.behaviour.hauler.js [hunk 2]
```diff
@@ -57,3 +57,37 @@ mod.nextAction = function(creep){
...
+
+            return function (amount) {
+                return amount >= min;
+            };
+        }
+    },
+    uncharging: {
+        name: `uncharging-${mod.name}`,
+        targetScore: function (creep) {
+            const isValidTarget = creep.getStrategyHandler(['uncharging'], 'isValidTarget', creep);
```
Comment (chobobobo) [inline creep.behaviour.hauler.js]: Is there a way for creep.getStrategyHandler to reference the caller, or does the creep have to be passed in with it, it seems redundant to me but I admit I don't fully understand how it works.

**Comment by karlthepagan** [comment]
Exactly right on your interpretation.

File: creep.behaviour.hauler.js (hunk 2 above)
Comment (karlthepagan) [inline creep.behaviour.hauler.js]: The strategy handler should not use the `this` context when it is defined. Any parameters should be passed in explicitly.

IMO this keeps it more clear for JS novices as well.

**Comment by whardeman** [comment]
My in-room haulers are completely ignoring source containers. They just don't even try to empty them.

**Comment by karlthepagan** [comment]
As a workaround uncomment the robbing line in remote worker behaviour.

File: creep.action.uncharging.js [hunk 3]
```diff
@@ -85,3 +57,48 @@ action.onAssignment = function(creep, target) {
...
+        } else if (target.structureType == 'container') {
+            const canWithdrawEnergy = creep.getStrategyHandler([action.name], 'canWithdrawEnergy', creep);
+            const withdrawAmount = creep.getStrategyHandler([action.name], 'withdrawAmount', target);
+            return canWithdrawEnergy(withdrawAmount);
+        }
+        return false;
+    };
+};
+action.defaultStrategy.canWithdrawEnergy = function(creep) {
+    const min = creep.body.carry * 25;
```
Comment (whardeman) [inline creep.action.uncharging.js]: `creep.body.carry * 25` = NaN

Should be `creep.data.body.carry`

Comment (whardeman) [inline creep.action.uncharging.js]: This is why haulers are ignoring containers.

**Comment by karlthepagan** [comment]
I'll get this one removed from Canary until I can fix it.

**Comment by karlthepagan** [comment]
I've been using uncharging + picking strategy merged with dev. I'm going to close this one and bug people to review the other when I update the branches.

**Comment by karlthepagan** [comment]
superseded by #685

**Pull request closed without merging**